diff --git a/dlls/kernelbase/file.c b/dlls/kernelbase/file.c
index f4f9d14518f..105fa6ab3e0 100644
--- a/dlls/kernelbase/file.c
+++ b/dlls/kernelbase/file.c
@@ -2917,6 +2917,18 @@ BOOL WINAPI DECLSPEC_HOTPATCH SetCurrentDirectoryA( LPCSTR dir )
 BOOL WINAPI DECLSPEC_HOTPATCH SetCurrentDirectoryW( LPCWSTR dir )
 {
     UNICODE_STRING dirW;
+    const char *sgi = getenv("SteamGameId");
+
+    if ((sgi != NULL) && (strcmp(sgi, "244210") == 0)) {
+        SIZE_T len = wcslen(dir);
+
+        if (len > 0 && dir[len - 1] == '.')
+        {
+            WCHAR *p = (WCHAR *)dir + len - 1;
+            *p = '\0';
+            FIXME("%s . fixed\n", debugstr_w(dir));
+        }
+    }
 
     RtlInitUnicodeString( &dirW, dir );
     return set_ntstatus( RtlSetCurrentDirectory_U( &dirW ));
